### PR TITLE
Numpy 1.18 support

### LINKED ISSUE #3409: test failure with upstream-dev
https://dev.azure.com/xarray/xarray/_build/results?buildId=1101&view=logs

```
=================================== FAILURES ===================================
_________________________ test_datetime_reduce[False] __________________________

dask = False

    @arm_xfail
    @pytest.mark.parametrize("dask", [False, True])
    def test_datetime_reduce(dask):
        time = np.array(pd.date_range("15/12/1999", periods=11))
        time[8:11] = np.nan
        da = DataArray(np.linspace(0, 365, num=11), dims="time", coords={"time": time})
    
        if dask and has_dask:
            chunks = {"time": 5}
            da = da.chunk(chunks)
    
        actual = da["time"].mean()
>       assert not pd.isnull(actual)
E       AssertionError: assert not True
E        +  where True = <function isna at 0x7f2475449ae8>(<xarray.DataArray 'time' ()>\narray('NaT', dtype='datetime64[ns]'))
E        +    where <function isna at 0x7f2475449ae8> = pd.isnull

xarray/tests/test_duck_array_ops.py:288: AssertionError
__________________________ test_datetime_reduce[True] __________________________

dask = True

    @arm_xfail
    @pytest.mark.parametrize("dask", [False, True])
    def test_datetime_reduce(dask):
        time = np.array(pd.date_range("15/12/1999", periods=11))
        time[8:11] = np.nan
        da = DataArray(np.linspace(0, 365, num=11), dims="time", coords={"time": time})
    
        if dask and has_dask:
            chunks = {"time": 5}
            da = da.chunk(chunks)
    
        actual = da["time"].mean()
>       assert not pd.isnull(actual)
E       AssertionError: assert not True
E        +  where True = <function isna at 0x7f2475449ae8>(<xarray.DataArray 'time' ()>\narray('NaT', dtype='datetime64[ns]'))
E        +    where <function isna at 0x7f2475449ae8> = pd.isnull

xarray/tests/test_duck_array_ops.py:288: AssertionError
=============================== warnings summary 
```

### PR DESCRIPTION
**Pull request by crusaderky**
Fix mean() and nanmean() for datetime64 arrays on numpy backend when upgrading from numpy 1.17 to 1.18.
All other nan-reductions on datetime64s were broken before and remain broken.
mean() on datetime64 and dask was broken before and remains broken.

 - [x] Closes #3409
 - [x] Passes `black . && mypy . && flake8`
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

### CODE REVIEW

**Comment by crusaderky** [comment]
xref #3434

**Comment by crusaderky** [comment]
Ready for review and merge

**Comment by mathause** [comment]
Given https://github.com/numpy/numpy/pull/14841 - is this worth the extra complexity?

Currently xarray is only broken wrt numpy master and there is a chance https://github.com/numpy/numpy/pull/14841 gets into 1.18 - then we could just use `np.nanmin`?

Also - is this the only place of `min` for datetimes? Should this also go into the operator dispatcher?

**Comment by dopplershift** [comment]
IMO, it's always best to release code that you *know* will work rather than rely on upstream to get something into the next release in order for you not to be broken. I say that both as a downstream user of xarray and a library maintainers.

**Comment by dcherian** [comment]
LGTM. I think we should merge so we can get a release out.

**Comment by crusaderky** [comment]
@dopplershift fully agree.

> is this the only place of min for datetimes? 

This PR *does not* fix the nan-reductions for datetime arrays with NaT.

It *very narrowly* fixes the use case of **mean() and nanmean() with a numpy backend**, so that they work on numpy 1.18 as they did on 1.17. As a matter of fact they never worked with dask, were never tested, and still won't work after the PR.

To emphasize: I am not introducing any new feature or fixing bugs on numpy 1.17; I'm just getting rid of regressions on the upgrade from 1.17 to 1.18.

Specifically, the PR fixes this failing test:
https://github.com/pydata/xarray/blob/aa876cfd6b3b97ee5028d089ec741d057e3ed688/xarray/tests/test_duck_array_ops.py#L276-L293

which was not well formulated - despite its name, it doesn't test all reductions, only mean(), and in fact never runs on dask (the index of a dask array is not chunked).
I have now rewritten it  for the sake of clarity.

*If* numpy/numpy#14841 makes into master before 1.18 is published, and *if* it actually delivers on making reductions work with NaT (I did not test it yet), I'll be happy to write a second PR and rework everything.
Even then, a switch of this sort will still be needed:

```python
def datetime_nanmin(array, **kwargs):
    if LooseVersion(numpy.__version__) < "1.18":
        return np.min(array, **kwargs)
    else:
        return np.nanmin(array, **kwargs)
```
The actual code will be in fact more complicated because you also need to think about dask.

**Comment by max-sixty** [comment]
Agree—let's merge?

**Comment by mathause** [comment]
Ok, thanks for the clarification.

**Comment by crusaderky** [comment]
Updated PR description:

Fix mean() and nanmean() for datetime64 arrays on numpy backend when upgrading from numpy 1.17 to 1.18.
All other nan-reductions on datetime64s were broken before and remain broken.
All reductions on datetime64s, including mean() and nanmean(), were broken on dask and remain broken.

**Comment by crusaderky** [comment]
I just realised there are no tests at all for ``datetime_to_numeric(offset=None) `` on a dask backend. I suspect I broke something - I need to look at it better.

**Comment by crusaderky** [comment]
``datetime_to_numeric(offset=None)`` never worked with dask to begin with - there's a big fat TODO in it about it. I rewrote most of the PR; now mean() works with dask.

Pre-empting an observation: *yes*, the nanmin function I wrote *could* be used to work around the shortcomings of both numpy and dask on datetime64. I don't think xarray should do it and IMHO the workaround currently implemented for mean() should never have been done to begin with.

Again, this PR only fixes the regression on numpy 1.18.

**Comment by crusaderky** [comment]
Ready for review and merge

File: xarray/core/duck_array_ops.py [hunk 1]
```diff
@@ -351,6 +351,32 @@ def f(values, axis=None, skipna=None, **kwargs):
...
+    Caveats that this function deals with:
+
+    - In numpy < 1.18, min() on datetime64 incorrectly ignores NaT
+    - numpy nanmin() don't work on datetime64 (all versions at the moment of writing)
+    - dask min() does not work on datetime64 (all versions at the moment of writing)
+    """
+    from .dataarray import DataArray
+    from .variable import Variable
+
+    if isinstance(array, (DataArray, Variable)):
```
Comment (dcherian) [inline xarray/core/duck_array_ops.py]: does this ever happen? All the other functions in this file work at the numpy, dask level (I think) so maybe this shouldn't be necessary.

Comment (crusaderky) [reply xarray/core/duck_array_ops.py]: Yes, I was surprised myself. I didn't investigate, but I agree it's not clean.

Comment (crusaderky) [reply xarray/core/duck_array_ops.py]: Fixed it

**Pull request merged**